src/config.c

```diff
@@ -0,0 +1,609 @@
+        cv_free(dir_buf);
+        cv_free(buf);
+        return NULL;
+    }
+    cv_free(dir_buf);
+    // returned path does not contain trailing '/'
+    return buf;
+}
+
+int config_set_file_path(const char *path)
+{
+    char *dirname = get_abs_dir(path);
+    if (dirname == NULL) {
+        return CORVUS_ERR;
+    }
+    char buf[CONFIG_FILE_PATH_SIZE + 1];
+    strncpy(buf, path, CONFIG_FILE_PATH_SIZE);
+    char *filename = basename(buf);
+    snprintf(config.config_file_path, CONFIG_FILE_PATH_SIZE, "%s/%s",
+        dirname, filename);
+    cv_free(dirname);
+    return CORVUS_OK;
+}
+
+int config_read(const char *filename)
+{
+    if (strlen(filename) > CONFIG_FILE_PATH_SIZE) {
+        fprintf(stderr, "Config file path is too long. Max length is %d.",
+            CONFIG_FILE_PATH_SIZE);
+        return CORVUS_ERR;
+    }
+
+    if (config_set_file_path(filename) == CORVUS_ERR) {
+        return CORVUS_ERR;
+    }
+
+    FILE *fp = fopen(filename, "r");
+    if (fp == NULL) {
+        // Maybe config.syslog hasn't been initialized yet
+        // so should not use LOG() here
+        fprintf(stderr, "config file: %s\n", strerror(errno));
+        return -1;
+    }
+    int len = 0;
+    struct cvstr line = cvstr_new(1024);
+    while ((len = read_line(&line, fp)) != -1) {
+        char name[len + 1], value[len + 1];
+        memset(name, 0, sizeof(name));
+        memset(value, 0, sizeof(value));
+        if (!parse_option(line.data, name, value)) {
+            continue;
+        }
+
+        if (config_add(name, value) == -1) {
+            cvstr_free(&line);
+            fclose(fp);
+            return CORVUS_ERR;
+        }
+    }
+    cvstr_free(&line);
+    fclose(fp);
+    return CORVUS_OK;
+}
+
+// Caller should free the returned vector.
+// Return an empty line array if fail to open or read old config file.
+struct vector get_curr_file_content()
+{
+    struct vector lines = vector_new();
+    FILE *fp = fopen(config.config_file_path, "r");
+    if (fp == NULL) {
+        LOG(ERROR, "Can't open config file: %s\n", strerror(errno));
+        return lines;
+    }
+
+    struct cvstr line = cvstr_new(128);
+    size_t len;
+    while ((len = read_line(&line, fp)) != -1) {
+        vector_push_back(&lines, cvstr_move(&line));
+        line = cvstr_new(128);
+    }
+    cvstr_free(&line);
+    fclose(fp);
+    return lines;
+}
+
+int gen_tmp_conf(const char *tmpfile)
+{
+    FILE *fp = fopen(tmpfile, "w");
+    if (fp == NULL) {
+        fclose(fp);
+        LOG(ERROR, "can't create tmp file: %s\n", strerror(errno));
+        return CORVUS_ERR;
+    }
+
+    int result = CORVUS_ERR;
+    // Create a brandnew config file if `lines` is empty
+    struct vector lines = get_curr_file_content();
+
+    const size_t OPTIONS_NUM = sizeof(CONFIG_OPTIONS) / sizeof(char*);
+    void *WRITTEN_TAG = (void*)1;
+    struct dict written_tags, options_map;
+    dict_init(&written_tags);
+    dict_init(&options_map);
```

Comment:
`dict_set` will not overwrite the same key. It just add the key again with another value.